### ISSUE TITLE
Error trying to update nan package from forge

### ISSUE THREAD

**Issue by kupiqu** [flathub/org.octave.Octave#80]
There are many errors as the ones below.

It could be that there is something missing in flatpak or that latest version of the package is not compatible with Octave 5.1 anymore, I cannot say.

```
/usr/lib/gcc/x86_64-unknown-linux-gnu/8.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: /tmp/oct-kThpMG/nan-3.4.3/src/kth_element.cpp:208: undefined reference to `mxFree'
collect2: error: ld returned 1 exit status
make: *** [Makefile:195: kth_element.mex] Error 1
make: *** Waiting for unfinished jobs....
```
```
c/str2array.cpp:209: undefined reference to `mxSetPi'
collect2: error: ld returned 1 exit status
make: *** [Makefile:195: str2array.mex] Error 1
```
```
/usr/lib/gcc/x86_64-unknown-linux-gnu/8.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: /tmp/oct-kThpMG/nan-3.4.3/src/histo_mex.cpp:233: undefined reference to `mexWarnMsgTxt'
collect2: error: ld returned 1 exit status
make: *** [Makefile:195: histo_mex.mex] Error 1
```
```
/usr/lib/gcc/x86_64-unknown-linux-gnu/8.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: /tmp/oct-UJxhIj.o: in function `mexFunction':
/tmp/oct-kThpMG/nan-3.4.3/src/covm_mex.cpp:837: undefined reference to `mexPrintf'
/usr/lib/gcc/x86_64-unknown-linux-gnu/8.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: /tmp/oct-kThpMG/nan-3.4.3/src/covm_mex.cpp:122: undefined reference to `mxGetPr'
collect2: error: ld returned 1 exit status
make: *** [Makefile:195: covm_mex.mex] Error 1
```
```
/usr/lib/gcc/x86_64-unknown-linux-gnu/8.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: /tmp/oct-kThpMG/nan-3.4.3/src/sumskipnan_mex.cpp:196: undefined reference to `mxFree'
collect2: error: ld returned 1 exit status
make: *** [Makefile:195: sumskipnan_mex.mex] Error 1
```
```
/usr/lib/gcc/x86_64-unknown-linux-gnu/8.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: /tmp/oct-kThpMG/nan-3.4.3/src/xptopen.cpp:577: undefined reference to `mxCreateStructMatrix'
collect2: error: ld returned 1 exit status
make: *** [Makefile:195: xptopen.mex] Error 1
make: Entering directory '/tmp/oct-kThpMG/nan-3.4.3/src'
/app/bin/mkoctfile-5.1.0 --mex -lgomp "histo_mex.cpp"
/app/bin/mkoctfile-5.1.0 --mex -lgomp "covm_mex.cpp"
/app/bin/mkoctfile-5.1.0 --mex -lgomp "kth_element.cpp"
/app/bin/mkoctfile-5.1.0 --mex -lgomp "sumskipnan_mex.cpp"
/app/bin/mkoctfile-5.1.0 --mex -lgomp "str2array.cpp"
g++ -fexceptions -fPIC -fno-omit-frame-pointer -pthread -c tron.cpp
g++ -fexceptions -fPIC -fno-omit-frame-pointer -pthread -c linear.cpp
/app/bin/mkoctfile-5.1.0 -o "svm_model_octave.o" -c "svm_model_matlab.c"
cc -fexceptions -fPIC -fno-omit-frame-pointer -pthread -c svm.cpp
/app/bin/mkoctfile-5.1.0 --mex -lgomp "xptopen.cpp"
make: Leaving directory '/tmp/oct-kThpMG/nan-3.4.3/src'

pkg: error running `make' for the nan package.
error: called from
    configure_make at line 99 column 9
    install at line 184 column 7
    pkg at line 441 column 9
    pkg at line 592 column 11
```

**Comment by mtmiller** [comment]
Confirmed. The root cause of these errors is the linker option `--no-undefined`, which is hard coded into the NaN package makefile [here](http://hg.code.sf.net/p/octave/NaN/file/3.4.3/src/Makefile.in#l133). I don't know why that option is used. And I don't know why that option is _not_ used when I install the package outside of Flatpak. The makefile is a bit hard to follow.

The same option will break any package with oct files in Octave 5, for example

    setenv LDFLAGS "-Wl,--no-undefined"
    pkg install -forge -verbose control
    ...
    /tmp/oct-PLwlX4/control-3.2.0/src/is_zp_vector.cc:41: undefined reference to `octave_value::octave_value(bool)'
    ...

This looks to me like a bug in the NaN package build rules. I would recommend removing the entire LDFLAGS definition from the makefile. But I'm not sure exactly what is causing the makefile to run different rules or why that option is only applied in one context but not another.

**Comment by mtmiller** [comment]
Investigated for a few minutes. The main difference in the Flatpak environment is that `LDFLAGS` is set to something, anything, in the user's environment.

If `LDFLAGS` is in the calling environment, then make propagates that variable into the environment of mkoctfile, etc.

I can reproduce this error outside of Flatpak by simply setting `LDFLAGS` to any valid option. For example

    $ export LDFLAGS=-v
    $ octave --eval "pkg install -forge nan"
    ... undefined reference to ...

So I think this should be reported to the NaN package and fixed there in a future update.

**Comment by kupiqu** [comment]
I see. Thanks for figuring this out!

Could you please report it to the maintainer of the `nan` package or, alternatively, let me know how can I do that?

I know how to report bugs against octave, but no idea about packages in octave forge :/

Is it at the same place?

**Comment by mtmiller** [comment]
Yes, reported as [Octave#57228](https://savannah.gnu.org/bugs/?57228). :heavy_check_mark:

**Comment by kupiqu** [comment]
Thank you very much

**Comment by mtmiller** [comment]
Btw, this also affected the Debian octave-nan package, and has been patched to fix it: https://salsa.debian.org/pkg-octave-team/octave-nan/blob/master/debian/patches/drop-no-undefined-ldflags.patch

I expect the only thing to do here is wait for a new nan package that does work, or patch it on your own.

**Comment by kupiqu** [comment]
I see. Thanks for telling

**Comment by kupiqu** [comment]
Unfortunately another issue has appeared:

```
xptopen.cpp: In function ‘void mexFunction(int, mxArray**, int, const mxArray**)’:
xptopen.cpp:510:9: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’, declared withattribute warn_unused_result [-Wunused-result]
    fread(data, bi[NS], M, fid);
    ~~~~~^~~~~~~~~~~~~~~~~~~~~~
svmpredict_mex.cpp: In function ‘int print_null(const char*, ...)’:
svmpredict_mex.cpp:55:36: warning: no return statement in function returning non-void [-Wreturn-type]
 int print_null(const char *s,...) {}
                                    ^
/usr/lib/gcc/x86_64-unknown-linux-gnu/8.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: cannot find -lblas
collect2: error: ld returned 1 exit status
make: *** [Makefile:184: train.mex] Error 1
make: *** Waiting for unfinished jobs....
make: Entering directory '/tmp/oct-1N17ji/nan-3.4.3/src'
/app/bin/mkoctfile-5.1.0 --mex -lgomp "histo_mex.cpp"
/app/bin/mkoctfile-5.1.0 --mex -lgomp "covm_mex.cpp"
/app/bin/mkoctfile-5.1.0 --mex -lgomp "kth_element.cpp"
/app/bin/mkoctfile-5.1.0 --mex -lgomp "sumskipnan_mex.cpp"
/app/bin/mkoctfile-5.1.0 --mex -lgomp "str2array.cpp"
g++ -fexceptions -fPIC -fno-omit-frame-pointer -pthread -c tron.cpp
g++ -fexceptions -fPIC -fno-omit-frame-pointer -pthread -c linear.cpp
/app/bin/mkoctfile-5.1.0 -o "svm_model_octave.o" -c "svm_model_matlab.c"
cc -fexceptions -fPIC -fno-omit-frame-pointer -pthread -c svm.cpp
/app/bin/mkoctfile-5.1.0 --mex -lgomp "xptopen.cpp"
/app/bin/mkoctfile-5.1.0 --mex -lgomp -lblas train.c tron.o linear.o linear_model_matlab.c
/app/bin/mkoctfile-5.1.0 --mex -lgomp svmtrain_mex.cpp svm_model_octave.o svm.o -o svmtrain_mex.mex
/app/bin/mkoctfile-5.1.0 --mex -lgomp svmpredict_mex.cpp svm_model_octave.o svm.o -o svmpredict_mex.mex
make: Leaving directory '/tmp/oct-1N17ji/nan-3.4.3/src'

pkg: error running `make' for the nan package.
error: called from
    configure_make at line 99 column 9
    install at line 184 column 7
    pkg at line 441 column 9
```

**Comment by mtmiller** [comment]
Yes, confirmed, I've filed this as [Octave#57263](https://savannah.gnu.org/bugs/?57263), along with a patch. With that patch and the last bug fixed, the nan package now installs.

**Comment by kupiqu** [comment]
Excellent, thank you!
